patches/typescript+4.4.4.patch

```diff
@@ -0,0 +1,194 @@
+diff --git a/node_modules/typescript/lib/typescript.js b/node_modules/typescript/lib/typescript.js
+index 323de6f..bb3e6b4 100644
+--- a/node_modules/typescript/lib/typescript.js
++++ b/node_modules/typescript/lib/typescript.js
+@@ -24,11 +24,11 @@ var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
+     return to.concat(ar || Array.prototype.slice.call(from));
+ };
+ var __assign = (this && this.__assign) || function () {
+-    __assign = Object.assign || function(t) {
```

Comment:
These changes to the `typescript` package were motivated by failures I was getting from LavaMoat when attempting to generate LavaMoat policy files for the build system (via `yarn lavamoat:auto`. That said, this commit is concerned with adding TypeScript to ESLint, not the build system, so why are these changes necessary?

Because, as I learned, the build system actually runs ESLint. As I've been going through the codebase, I've occasionally come across sections of code that are surrounded with `///: BEGIN:ONLY_INCLUDE_IN` and `///: END:ONLY_INCLUDE_IN` (for example, [`account-menu.component.js`](https://github.com/MetaMask/metamask-extension/blob/2b010710aec96f3bf74bb07d7c56cfa2c1c59d7a/ui/components/app/account-menu/account-menu.component.js#L15-L17). I've always wondered about them, and now I know that they get removed [during a step in the build process](https://github.com/MetaMask/metamask-extension/blob/2b010710aec96f3bf74bb07d7c56cfa2c1c59d7a/development/build/transforms/remove-fenced-code.js). After this transform is applied to a file, [that file is linted](https://github.com/MetaMask/metamask-extension/blob/2b010710aec96f3bf74bb07d7c56cfa2c1c59d7a/development/build/transforms/utils.js#L61).

So, when running `yarn lavamoat:auto`, LavaMoat sees these files and follows the path to `eslint.config.js`, then follows the path to the `typescript` package. Why does it trip up on this file in particular?

One of the things that LavaMoat, or more accurately the `lockdown` function in SES, does is to protect against unwanted and unexpected modifications of global objects, such as `Object.prototype`. Admittedly I am unable to explain much beyond this, but you can learn more about this function [here](https://github.com/endojs/endo/blob/master/packages/ses/README.md#lockdown). The `lockdown` function, as it uses `Object.freeze` to lock down objects, suffers from a ["mistake"](https://web.archive.org/web/20141230041441/http://wiki.ecmascript.org/doku.php?id=strawman:fixing_override_mistake) in the ECMAScript spec where a property inherited from a frozen object is itself frozen. Plain objects inherit `constructor` from `Object.prototype`, so if `Object.prototype` is frozen, this code will not work:

``` javascript
const foo = {};
foo["constructor"] = "bar";
```

But that is what is happening in this file. The `ts.textToKeywordObj` object maps names of keywords in TypeScript to token ids that represent those keywords (defined elsewhere), and one of the keywords in TypeScript is `constructor`. For some reason, whatever ended up producing this file (probably TypeScript itself) defined this object essentially like this:

``` javascript
ts.textToKeywordObj = (
  _a = { /* ... */ }, 
  _a["constructor"] = 133,
  // ...
  _a
);
```

That doesn't work, so one of the changes made here is to clean up the definition of this object.

The other change is to the `__assign` function, which you can see here. One of the ways that `__assign` is used is to make a `textToToken` variable on line 9904. This looks something like:

``` javascript
var textToToken = __assign(
  {},
  ts.textToKeywordObj,
  {
    "{": 18 /* OpenBraceToken */,
    "}": 19 /* CloseBraceToken */,
    // ...
  }
)
```

As you can see here, `__assign` defaults to `Object.assign` here if it exists, or otherwise falls back to a polyfill of `Object.assign`. Whether or not the polyfill is used, `__assign` creates a new object by iterating over the given objects and manually copying their keys and values to the new object by setting them directly on the new object. This will attempt to set `constructor` on the new object again, which as we've seen, is not allowed. So this change fixes `__assign` by using `Object.defineProperty` to set properties instead of direct assignment.